src/settings/settings_manager.cpp

```diff
@@ -0,0 +1,227 @@
+#include <gflags/gflags.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "common/macros.h"
+#include "main/db_main.h"
+#include "settings/settings_manager.h"
+#include "type/transient_value_factory.h"
+
+#define __SETTING_GFLAGS_DECLARE__     // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_GFLAGS_DECLARE__      // NOLINT
+
+namespace terrier::settings {
+
+using Index = catalog::SettingsTableColumn;
+using ValueFactory = type::TransientValueFactory;
+using ValuePeeker = type::TransientValuePeeker;
+using ActionContext = common::ActionContext;
+using ActionState = common::ActionState;
+
+// Used for building temporary transactions
+void EmptyCallback(void * /*unused*/) {}
+
+SettingsManager::SettingsManager(DBMain *db, catalog::Catalog *catalog, transaction::TransactionManager *txn_manager)
+    : db_(db), settings_handle_(catalog->GetSettingsHandle()), txn_manager_(txn_manager) {
+  ValidateParams();
+  InitializeCatalog();
+}
+
+void SettingsManager::ValidateParams() {
+  // This will expand to invoke settings_manager::DefineSetting on
+  // all of the settings defined in settings.h.
+  // Example:
+  //   ValidateSetting(Param::port, type::TransientValueFactory::GetInteger(1024),
+  //                  type::TransientValueFactory::GetInteger(65535));
+
+#define __SETTING_VALIDATE__           // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_VALIDATE__            // NOLINT
+}
+
+void SettingsManager::ValidateSetting(Param param, const type::TransientValue &min_value,
+                                      const type::TransientValue &max_value) {
+  const ParamInfo &info = db_->param_map_.find(param)->second;
+  if (!ValidateValue(info.value, min_value, max_value)) {
+    SETTINGS_LOG_ERROR(
+        "Value given for \"{}"
+        "\" is not in its min-max bounds",
+        info.name);
+    throw SETTINGS_EXCEPTION("Invalid setting value");
+  }
+}
+
+void SettingsManager::InitializeCatalog() {
+  auto txn = txn_manager_->BeginTransaction();
+  auto column_num = catalog::SettingsHandle::schema_cols_.size();
+
+  for (const auto &pair : db_->param_map_) {
+    const Param param = pair.first;
```

Comment:
You can use C++17 structured bindings here if you wish. This can simplify handling of data structures such as `std::pair`.

Reply:
Thank you for the reminder :)